warcprox/dedup.py

```diff
@@ -412,17 +415,30 @@ def _filter_and_bucketize(self, batch):
         '''
         buckets = collections.defaultdict(list)
         discards = []
+        # for duplicate checks, see https://webarchive.jira.com/browse/WT-31
+        hash_plus_urls = set()
         for recorded_url in batch:
+            if recorded_url.payload_digest:
+                hash_plus_url = b''.join((warcprox.digest_str(
+                            recorded_url.payload_digest, self.options.base32), recorded_url.url))
             if (recorded_url.response_recorder
                     and recorded_url.payload_digest
-                    and self.trough_dedup_db.should_dedup(recorded_url)):
+                    and self.trough_dedup_db.should_dedup(recorded_url)
+                    and hash_plus_url not in hash_plus_urls):
+                hash_plus_urls.add(hash_plus_url)
                 if (recorded_url.warcprox_meta
                         and 'dedup-buckets' in recorded_url.warcprox_meta):
                     for bucket, bucket_mode in recorded_url.warcprox_meta["dedup-buckets"].items():
                         buckets[bucket].append(recorded_url)
                 else:
                     buckets['__unspecified__'].append(recorded_url)
             else:
+                if hash_plus_url in hash_plus_urls:
```

Comment:
You use `hash_plus_url` here but if `recorded_url.payload_digest` doesn't exist, it won't be defined at all.

Maybe its better to say:
```
if recorded_url.payload_digest:
    hash_plus_url = b''.join((warcprox.digest_str(
                            recorded_url.payload_digest, self.options.base32), recorded_url.url))
else:
    continue
```
in the beginning of the loop.

Reply:
Actually, it may be even better to just say:
```
if not recorded_url.payload_digest:
    continue
```
in the beginning of the loop and avoid checking it 2 times later.